infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msghandlers/LogReplicationServer.java

```diff
@@ -296,17 +298,27 @@ private void handleMetadataRequest(RequestMsg request, ResponseMsg res,
     }
 
     /**
-     * Given a leadership request message, send back a
-     * response indicating our current leadership status.
+     * Given a leadership request message, send back a response indicating our current leadership status.
+     * Since the client registration happens on the SOURCE, leadership query msg also serves as a vehicle to notify the
+     * SINK about the new sessions. Response to the leadership request msg is sent only after processing the session
+     * information in the msg header.
      *
      * @param request the leadership request message
      * @param router  router used for sending back the response
      */
     @LogReplicationRequestHandler(requestType = LR_LEADERSHIP_QUERY)
-    private void  handleLeadershipQuery(RequestMsg request, ResponseMsg res,
+    private void handleLeadershipQuery(RequestMsg request, ResponseMsg res,
                                                      @Nonnull IClientServerRouter router) {
         log.debug("Log Replication Query Leadership Request received by Server.");
         HeaderMsg responseHeader = getHeaderMsg(request.getHeader());
+        // if session in the leadership query is not known to the local cluster, create and process the new session information
+        if (replicationContext.getIsLeader().get()) {
+            LogReplicationSession session = getSession(request);
+            if (!sessionManager.getIncomingSessions().contains(session)) {
+                this.sessionManager.addNotifiedSessionFromSource(session);
+                createSinkManager(session);
+            }
+        }
```

Comment:
yet to examine all the path, but I think this new workflow would cause some issues around the onConnectionDown/Up path, basically after connection resume (or any kind of node restart), the Sink side could have lost the session info and we should always go back to leadership verify phase for the sessions to be created on Sink side.

Is it already covered somewhere?

Reply:
Yes, its already covered. 
Connection Down/UP events are always followed by leadership query. 
Incase the connection down is transient and the CONNECTION_DOWN event is not called, the sink retains the session info. The session info once received is only cleared on session-stop or LR shut down.